client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -272,7 +276,31 @@ public void wrongHost(WrongHost wrongHost) {
 
         @Override
         public void segmentIsSealed(SegmentIsSealed segmentIsSealed) {
-            state.failConnection(new SegmentSealedException());
+            log.trace("Received SegmentSealed {}", segmentIsSealed);
+            if (state.segmentSealed.compareAndSet(false, true)) {
+                CompletableFuture.<Void>supplyAsync(() -> {
+                    /*
+                      Invariants for segment sealed:
+                      * SegmentSealed callback and write will not run concurrently.
```

Comment:
The issue is that 
- getUnckedEvents() is not read after the callback acquires the lock.
- close connection of the sealed segment  should be performed inside the lock.

As per our discussion have updated the code as follows
1. Change getUnackedEvents() to seal (). 
2. Ensure segmentOutputStream.write() still invokes getConnection() to handle connection failures.
3. Ensure the segmentSelector invokes `writer.seal()` (this ensures we get unacked events within the readlock and we are certain that there are no other missing events)